### ISSUE TITLE
Feature Request: for update

### ISSUE THREAD

**Issue by xserve98** [ent/ent#358]
hello . 
 Is it possible to have pessimistic Lock functionality similar ？
 [https://dev.mysql.com/doc/refman/8.0/en/innodb-locking-reads.html](https://dev.mysql.com/doc/refman/8.0/en/innodb-locking-reads.html)

**Comment by mingrammer** [comment]
I need it too for postresql

**Comment by EricMasonFa** [comment]
Let's discuss if we can achive it like this:

In generated config.go:

```go
type config struct {
        // If we need to locking on tx or not
        withLock bool
	// driver used for executing database requests.
	driver dialect.Driver
	// debug enable a debug logging.
	debug bool
	// log used for logging on debug mode.
	log func(...interface{})
	// hooks to execute on mutations.
	hooks *hooks
}
```

In generated tx.go:
```go
func (tx *Tx) WithLock() *Client {
	tx.client.config.withLock = true
	return tx.client
}
```

Then we can pass that field to `sql.Selector` (we can ignore it if the driver does not support locking):
```go
        if s.withLock == true {
		b.WriteString(" FOR SHARE")
	}
```

Then we can lock/unlock like:
```go
func WithLockedTx(ctx context.Context, client *ent.Client, fn func(tx *ent.Tx) error) (tx, error) {
	tx, err := client.Tx(ctx).WithLock()
	if err != nil {
		return nil, err
	}
	tx = tx.WithLock()

	if err := fn(tx); err != nil {
		if rerr := tx.Rollback(); rerr != nil {
			err = errors.Wrapf(err, "rolling back transaction: %v", rerr)
		}
		return nil, err
	}
	
	return tx, nil
}

func UnlockTx(tx *ent.Tx) error {
         defer func() {
		if v := recover(); v != nil {
			tx.Rollback()
			panic(v)
		}
	}()

        if err := tx.Commit(); err != nil {
		return errors.Wrapf(err, "committing transaction: %v", err)
	}

        return nil
}

tx, err := WithLockedTx(ctx, client, func(tx *ent.Tx) error {
	return Gen(ctx, tx.Client())
})

err != nil {
	log.Fatal(err)
}

// do some stuff while tx result is locked

err := UnlockTx(tx)
err != nil {
	log.Fatal(err)
}
```

If the query is out of transaction but `withLock` is set to `true` then the lock would immediately be released by the driver.

**Comment by errorhandler** [comment]
Just chiming in here as this is something we need for our use of Ent. @EricMasonFa's suggestion looks really good. However, as far I as can see, that only supports locking *all* `SELECT`s in a transaction not just some.

One way I can see is to add a new method to the generated `{Object}Query` struct that allows a query to opt to locking, for example:

```go
type PetQuery struct {
	// ...
	sql  *sql.Selector
	// ...
}

func (pq *PetQuery) WithLock() *PetQuery {
	sql.withLock = true
	return pq
}
```

I understand that this pollutes the generated code but increases the increase in flexibility is worth it my opinion.

**Comment by EricMasonFa** [comment]
> Just chiming in here as this is something we need for our use of Ent. @EricMasonFa's suggestion looks really good. However, as far I as can see, that only supports locking _all_ `SELECT`s in a transaction not just some.
> 
> One way I can see is to add a new method to the generated `{Object}Query` struct that allows a query to opt to locking, for example:
> 
> ```go
> type PetQuery struct {
> 	// ...
> 	sql  *sql.Selector
> 	// ...
> }
> 
> func (pq *PetQuery) WithLock() *PetQuery {
> 	sql.withLock = true
> 	return pq
> }
> ```
> 

I really like this. In this way, we can have multiple types of locks:

```go
func (pq *PetQuery) LockForUpdate() *PetQuery {
	sql.withLock = Pet.updateLock
	return pq
}

func (pq *PetQuery) LockForShare() *PetQuery {
	sql.withLock = Pet.shareLock
	return pq
}
```

And maybe even table lock:

```go
func (pq *PetClient) LockTable() *PetLock {
	return &PetLock{config: c.config}
}
```

**Comment by rubensayshi** [comment]
really like the suggestions here and I do think it should be on the `{Object}Query` since you don't always want to lock all.

I guess a question left to answer is, should the `With{Edge}()` queries also inherit the chosen lock strategy?

I can put in the work to implement this if there's somewhat of an agreement on the direction and support from the maintainers that they'd like to see this added. @a8m?

**Comment by ryotarai** [comment]
FYI: I implemented a wrapper of Driver as a workaround: https://gist.github.com/ryotarai/4b4fba383d7cfa33566c3959cdad78c6

**Comment by a8m** [comment]
Hey all, the issue was resolved with #1755. The documentation for this feature exists in [entgo.io website](https://entgo.io/docs/feature-flags/#row-level-locks), and an extensive example/blog-post will be posted in the upcoming days. Thanks for all the suggestions here. It's appreciated 🙏 

cc @rotemtam
